### DISCUSSION TITLE
Debugging the Generation of pure_dart

### DISCUSSION THREAD

**Discussion by coder0xff** [fzyzcjy/flutter_rust_bridge#1092] [Q&A] [answered]
I've made some changes to the parser. Now rust-analyzer is erroring out, showing a stack trace with some of my changes.

```
[ERROR rust_analyzer::lsp_utils] failed to run build scripts:
error: failed to run custom build command for `flutter_rust_bridge_example_single_block_test v1.0.0 (/home/coder0xff/Dropbox/Documents/Projects/Community/flutter_rust_bridge/frb_example/pure_dart/rust)`

... truncated ...

     0: rust_begin_unwind
               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
     1: core::panicking::panic_fmt
               at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
     2: lib_flutter_rust_bridge_codegen::parser::Parser::parse_function::{{closure}}
               at /home/coder0xff/Dropbox/Documents/Projects/Community/flutter_rust_bridge/frb_codegen/src/parser/mod.rs:243:25

```

I'm guessing that generating pure_dart serves as a test case, though I'm not sure why that would happen in rust-analyzer. I can't figure out how that process gets triggered. Is there something in the Cargo.toml that configures it as a test?

I found `frb_example/pure_dart/rust/build.rs` which seems like the entry point, because it's calling frb_codegen. However, I can't even build it, because when I run `cargo build` in the `pure_dart/rust directory` I just get the same failure as above, crashing inside my code. I'd like to use VS Code to launch and debug this, but I can't figure out what VS Code should even be invoking. (Or what program to start LLDB with.)

**Comment by fzyzcjy** [comment]
Looks like it is called in flutter_rust_bridge/frb_codegen/src/parser/mod.rs:243:25. If you show more stack traces you can see wher it is called

**Comment by coder0xff** [comment, marked as answer]
After putting an infinite loop in my code to make sure it doesn't exit, I ran `top` and saw that `build-script-build` was pinned at 100% CPU usage. I tried to attach the debugger to it, but it didn't work for unknown reasons. I searched the internet for build-script-build and landed on "build.rs" being a special file.

The answer I was looking for was here:
[Build Scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html#build-scripts) - The Cargo Book

**Reply by coder0xff** [reply]
I was able to debug in VS Code.

			{
				"type": "lldb",
				"request": "attach",
				"name": "cargo build",
				"waitFor": true,
				"program": "build-script-build"
			}

This launch configuration will make the debugger wait for `build-script-build` to start and attach when it sees it. I then run `cargo build` separately, and the debugger sees it. To make sure the debugger has plenty of time to attach before anything important happens, I also added this function:

    fn wait_for_debugger() {
        unsafe {
            let mut waiting = true;
            while core::ptr::read_volatile(&waiting) {}
        }
    }

I can call this function anywhere important to make sure the debugger is attached before the important thing happens. One the debugger is attached enter `set waiting = 0` and the program continue.
